util/merkle/src/triedbmut.rs

```diff
@@ -0,0 +1,539 @@
+// Copyright 2018 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use ccrypto::{blake256, BLAKE_NULL_RLP};
+use hashdb::DBValue;
+use hashdb::HashDB;
+use primitives::H256;
+
+use super::node::Node as RlpNode;
+use super::{TrieError, TrieMut};
+use nibbleslice::NibbleSlice;
+
+
+fn empty_child() -> [Option<H256>; 16] {
+    [None; 16]
+}
+
+pub struct TrieDBMut<'a> {
+    db: &'a mut HashDB,
+    root: &'a mut H256,
+}
+
+impl<'a> TrieDBMut<'a> {
+    /// Create a new trie with backing database `db` and empty `root`.
+    pub fn new(db: &'a mut HashDB, root: &'a mut H256) -> Self {
+        *root = BLAKE_NULL_RLP;
+
+        TrieDBMut {
+            db,
+            root,
+        }
+    }
+
+    /// Create a new trie with the backing database `db` and `root.
+    /// Returns an error if `root` does not exist.
+    pub fn from_existing(db: &'a mut HashDB, root: &'a mut H256) -> super::Result<Self> {
+        if !db.contains(root) {
+            return Err(Box::new(TrieError::InvalidStateRoot(*root)))
+        }
+
+        Ok(TrieDBMut {
+            db,
+            root,
+        })
+    }
+
+    /// Insert auxiliary
+    fn insert_aux(
+        &mut self,
+        path: NibbleSlice,
+        v_: DBValue,
+        cur_node_hash: Option<H256>,
+        old_val: &mut Option<DBValue>,
+    ) -> super::Result<H256> {
+        match cur_node_hash {
+            Some(hash) => {
+                let node_rlp = self.db.get(&hash).ok_or_else(|| Box::new(TrieError::IncompleteDatabase(hash)))?;
+
+                match RlpNode::decoded(&node_rlp[..]) {
+                    Some(RlpNode::Leaf(partial, v)) => {
+                        // Renew the Leaf
+                        if partial == path {
+                            let node = RlpNode::Leaf(path, v_);
+                            let node_rlp = RlpNode::encoded(Some(node), None);
+                            let hash = self.db.insert(&node_rlp[..]);
+
+                            *old_val = Some(v);
+
+                            self.db.remove(&cur_node_hash.expect("Can not change Leaf"));
```

Comment:
Node removal is handled outside of `TrieDB`. You must not remove any nodes while insertion.